data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderBase.java

```diff
@@ -15,12 +15,42 @@
 
 public abstract class TokenProviderBase {
     protected static final String ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN = "Error acquiring ApplicationAccessToken";
+    private static final String ERROR_INVALID_SERVICE_RESOURCE_URL = "Error determining scope due to invalid Kusto Service Resource URL";
     protected final Logger logger = LoggerFactory.getLogger(getClass());
     protected final String clusterUrl;
+    protected CloudInfo cloudInfo = null;
 
     TokenProviderBase(@NotNull String clusterUrl) throws URISyntaxException {
         this.clusterUrl = UriUtils.setPathForUri(clusterUrl, "");
     }
 
+    protected void initializeCloudInfo() throws DataServiceException {
+        if (cloudInfo == null) {

```

Comment:
!= ?

Reply:
Good catch. When I finally got around to running the E2E tests, this bug caused them **all** to fail :-D